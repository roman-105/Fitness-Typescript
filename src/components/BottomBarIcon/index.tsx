import React from 'react';
import { HomeIcon, ExploreIcon, ProgressIcon, CoachIcon, MyClubIcon } from '../Icon';
import Typography from '../Typography/Typography';
import styles from '../Typography/typography-styles';
import formatMessage from 'format-message';
import { Routes } from '../../router/routes';

const lookup: { [index: string]: (color: string) => JSX.Element } = {
  [Routes.Home]: (color: string) => <HomeIcon fill={color} />,
  [Routes.Programs]: (color: string) => <ExploreIcon fill={color} />,
  [Routes.Progress]: (color: string) => <ProgressIcon fill={color} />,
  [Routes.Coach]: (color: string) => <CoachIcon fill={color} />,
  [Routes.Club]: (color: string) => <MyClubIcon fill={color} />
};

export function bottomBarIcon(name: Routes) {
  return function ({ color }: { color: string }): JSX.Element {
    return lookup[name](color);
  };
}

export function bottomBarLabel(name: Routes) {
  const tabNameLookup: { [index: string]: string } = {
    [Routes.Home]: formatMessage('Home'),
    [Routes.Programs]: formatMessage('Explore'),
    [Routes.Progress]: formatMessage('Progress'),
    [Routes.Coach]: formatMessage('Coach'),
    [Routes.Club]: formatMessage('Clubs')
  };

  return function ({ color }: { color: string }): JSX.Element {
    return (
      <Typography
        style={{ color, ...styles.label }}
        fontFamily="regular"
        fontSize={11}
        lineHeight={12}
      >
        {tabNameLookup[name]}
      </Typography>
    );
  };
}
