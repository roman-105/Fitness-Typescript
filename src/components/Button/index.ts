export { default as BFButton } from './BFButton';
export { default as BFSecondaryButton } from './BFSecondaryButton';
export { default as BFSelectorButton } from './selector/BFSelectorButton';
export { default as BFLightButton } from './BFLightButton';
