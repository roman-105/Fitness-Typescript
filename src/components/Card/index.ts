export { default as MediaCard } from './Media/MediaCard';
export { default as SlashedCard } from './SlashedCard/SlashedCard';
