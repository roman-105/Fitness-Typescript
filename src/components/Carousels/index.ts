export { default as MediaCardCarousel } from './MediaCardCarousel';
import { MediaCardCarouselProps as mediaprops } from './MediaCardCarousel';
export { default as ImageCarousel } from './ImageCarousel';
import { ImageCarouselProps as imageprops } from './ImageCarousel';
export type MediaCardCarouselProps = mediaprops;
export type ImageCarouselProps = imageprops;
