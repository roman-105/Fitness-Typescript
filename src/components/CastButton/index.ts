export { default as BFCastButton } from './BFCastButton';
