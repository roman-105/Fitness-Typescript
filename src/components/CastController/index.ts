export { default as CastController } from './CastController';
