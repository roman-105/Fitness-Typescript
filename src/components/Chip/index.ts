export { default as ChipList } from './ChipList';
export { default as ChipItem } from './ChipItem';
import { Item } from './ChipList';
export type ChipItemType = Item;
