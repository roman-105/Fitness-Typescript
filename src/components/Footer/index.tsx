import React, { ReactElement } from 'react';
import { View } from 'react-native';
import styles from './footer-styles';

import Typography from '../Typography';
import { TouchableOpacity } from 'react-native-gesture-handler';
import { ExternalIcon, ArrowRightIcon } from '../Icon/';
import { BFIScreenBottomButton } from '../../store/models/screens';
import { SvgUri } from 'react-native-svg';
import { Routes } from '../../router/routes';
import formatMessage from 'format-message';
import theme from '../../theme';

interface itemProps {
  left: () => JSX.Element;
  text: string;
  right: () => JSX.Element;
  onPress: () => void;
}

interface FooterProps {
  children?: JSX.Element[];
}

function handlePress(navigation: any, type: 'External' | 'Internal', dispatch: any, link?: string) {
  if (link) {
    if (type === 'External') {
      navigation.navigate(Routes.Webview, { deeplink: link });
    } else {
      if (Object.values(Routes).includes(link.toUpperCase() as Routes)) {
        navigation.navigate(link.toUpperCase());
      } else {
        dispatch.snackbarModel.report({
          message: formatMessage('no screen {link} found', { link })
        });
      }
    }
  } else {
    // throw a snackbar with no content found.
    dispatch.snackbarModel.report({ message: formatMessage('no link found') });
  }
}

function Footer({ children }: FooterProps): JSX.Element | null {
  if (children && children.length > 0) {
    return (
      <View style={styles.footerContainer}>
        <Typography fontSize={14} fontFamily="trebleHeavy">
          Quick services
        </Typography>
        <View style={styles.childrenContainer}>{children}</View>
      </View>
    );
  }
  return null;
}

function FooterItem({ left, text, right, onPress }: itemProps): JSX.Element {
  return (
    <TouchableOpacity onPress={onPress}>
      <View style={styles.itemContainer}>
        <View style={styles.itemHeader}>
          {left()}
          <Typography fontSize={14} lineHeight={14} style={styles.text} fontFamily="medium">
            {text}
          </Typography>
        </View>
        {right()}
      </View>
    </TouchableOpacity>
  );
}

interface FooterFromConfigProps {
  config?: BFIScreenBottomButton[];
  dispatch: any;
  navigation: any;
}

export function FooterFromConfig(props: FooterFromConfigProps): ReactElement | null {
  if (!props.config) {
    return null;
  }
  return (
    <Footer>
      {/* show 3 buttons maximum */}
      {props.config.slice(0, 3).map((button, index) => (
        <FooterItem
          key={index}
          text={button.label}
          left={() => <SvgUri uri={`http:${button.icon}`} />}
          right={() =>
            button.type === 'External' ? (
              <ExternalIcon stroke={theme.colors.black} />
            ) : (
              <ArrowRightIcon fill={theme.colors.black} />
            )
          }
          onPress={() => handlePress(props.navigation, button.type, props.dispatch, button.link)}
        />
      ))}
    </Footer>
  );
}
