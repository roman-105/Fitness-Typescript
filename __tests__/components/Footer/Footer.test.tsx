// @ts-nocheck
import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { FooterFromConfig } from '../../../src/components/Footer';

jest.mock('react-native-svg', () => {
  return { __esModule: true, SvgUri: () => <></> };
});

const footerConfigMock = [
  {
    label: 'Hello, Im a mock #1',
    type: 'External',
    link: 'google.com',
    icon: 'im a mock :)'
  },
  {
    label: 'Hello, Im a mock #2',
    type: 'Internal',
    link: 'QR',
    icon: 'im a mock :)'
  },
  {
    label: 'Hello, Im a mock #3',
    type: 'Internal',
    link: 'PROFILE',
    icon: 'im a mock :)'
  },
  {
    label: 'Hello, Im a mock #4',
    type: 'Internal',
    link: 'google.com',
    icon: 'im a mock :)'
  }
];

const InvalidfooterConfigMock = [
  {
    label: 'Hello, Im a mock #1',
    type: 'External',
    link: null,
    icon: 'im a mock :)'
  },
  {
    label: 'Hello, Im a mock #2',
    type: 'Internal',
    link: 'route non existant',
    icon: 'im a mock :)'
  }
];

describe('Footer', () => {
  it('renders the correct content', async () => {
    const { toJSON } = render(
      <FooterFromConfig
        config={footerConfigMock}
        navigation={{ navigate: jest.fn }}
        dispatch={{ snackbarModel: { report: jest.fn } }}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('taps the right button / check internal external behaviour', async () => {
    const navigationFn = jest.fn();

    const { getByText } = render(
      <FooterFromConfig
        config={footerConfigMock}
        navigation={{ navigate: navigationFn }}
        dispatch={{ snackbarModel: { report: jest.fn } }}
      />
    );

    const externalLink = getByText('Hello, Im a mock #1');
    const internalLink1 = getByText('Hello, Im a mock #2');
    const internalLink2 = getByText('Hello, Im a mock #3');

    await fireEvent.press(externalLink);
    expect(navigationFn).toHaveBeenLastCalledWith(
      'WEBVIEW',
      expect.objectContaining({ deeplink: expect.stringContaining('google.com') })
    );
    await fireEvent.press(internalLink1);
    expect(navigationFn).toHaveBeenLastCalledWith('QR');
    await fireEvent.press(internalLink2);
    expect(navigationFn).toHaveBeenLastCalledWith('PROFILE');
  });

  it('taps the right button / reports errors', async () => {
    const dispatch = jest.fn();

    const { getByText } = render(
      <FooterFromConfig
        config={InvalidfooterConfigMock}
        navigation={{ navigate: jest.fn }}
        dispatch={{ snackbarModel: { report: dispatch } }}
      />
    );

    const externalLink = getByText('Hello, Im a mock #1');
    const internalLink1 = getByText('Hello, Im a mock #2');

    await fireEvent.press(externalLink);
    expect(dispatch).toHaveBeenLastCalledWith(
      expect.objectContaining({ message: expect.stringContaining('no link') })
    );
    await fireEvent.press(internalLink1);
    expect(dispatch).toHaveBeenLastCalledWith(
      expect.objectContaining({ message: expect.stringContaining('no screen') })
    );
  });

  it('dosent renders when no childrens are present', () => {
    const { toJSON } = render(
      <FooterFromConfig
        navigation={{ navigate: jest.fn }}
        dispatch={{ snackbarModel: { report: jest.fn } }}
      />
    );
    expect(toJSON()).toBeNull();
  });

  it('renders from a config and no more than 3 items', () => {
    const { queryByText } = render(
      <FooterFromConfig
        config={footerConfigMock}
        navigation={{ navigate: jest.fn }}
        dispatch={{ snackbarModel: { report: jest.fn } }}
      />
    );

    const additionalButton = queryByText('Hello, Im a mock #4');
    expect(additionalButton).toBeNull();
  });
});
