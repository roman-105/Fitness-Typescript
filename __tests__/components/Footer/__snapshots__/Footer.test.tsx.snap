// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer renders the correct content 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "paddingHorizontal": 20,
      "paddingVertical": 16,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "HeadingProTreble-Heavy",
        },
        undefined,
        Object {
          "fontSize": 14,
        },
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Quick services
  </Text>
  <View
    style={
      Object {
        "marginTop": 14,
      }
    }
  >
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
    >
      <View
        accessible={true}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderColor": "rgba(49,49,54,0.32)",
              "borderRadius": 2,
              "borderWidth": 1,
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 8,
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "ArgumentumMedium",
                  },
                  Object {
                    "marginLeft": 8,
                  },
                  Object {
                    "fontSize": 14,
                  },
                  Object {
                    "lineHeight": 14,
                  },
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              Hello, Im a mock #1
            </Text>
          </View>
          <test-file-stub
            stroke="#000000"
          />
        </View>
      </View>
    </RNGestureHandlerButton>
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
    >
      <View
        accessible={true}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderColor": "rgba(49,49,54,0.32)",
              "borderRadius": 2,
              "borderWidth": 1,
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 8,
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "ArgumentumMedium",
                  },
                  Object {
                    "marginLeft": 8,
                  },
                  Object {
                    "fontSize": 14,
                  },
                  Object {
                    "lineHeight": 14,
                  },
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              Hello, Im a mock #2
            </Text>
          </View>
          <test-file-stub
            fill="#000000"
          />
        </View>
      </View>
    </RNGestureHandlerButton>
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
    >
      <View
        accessible={true}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderColor": "rgba(49,49,54,0.32)",
              "borderRadius": 2,
              "borderWidth": 1,
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 8,
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "ArgumentumMedium",
                  },
                  Object {
                    "marginLeft": 8,
                  },
                  Object {
                    "fontSize": 14,
                  },
                  Object {
                    "lineHeight": 14,
                  },
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              Hello, Im a mock #3
            </Text>
          </View>
          <test-file-stub
            fill="#000000"
          />
        </View>
      </View>
    </RNGestureHandlerButton>
  </View>
</View>
`;
